### PR TITLE
Windows dns config test

### LINKED ISSUE #73414: Need Windows DNS configuration tests
We need to add Windows-specific DNS tests to cover the following fields:

  - Test cases needed for `dnsPolicy`: Default, ClusterFirst, None
  - Test cases needed for `dnsConfig`
  - Test cases needed for `hostname`
  - Test cases needed for `subdomain`

These need to run on Windows containers because the domain resolution policy is different, so different results need to be checked. Windows assumes that any name with a `.` in it is a FQDN, not a PQDN. 

There is an [open PR](https://github.com/kubernetes/kubernetes/pull/72729) to cover simple PQDN resolution, but they don't cover the Pod DNS configurations above. 

/sig windows
/milestone v1.14

### PR DESCRIPTION
**Pull request by lee0c**
**What type of PR is this?**

/sig windows
/kind feature

**What this PR does / why we need it**:

Adds 2 windows DNS tests. Hostname & subdomain test should be working, second test tests functionality that is currently broken as reported by #74541 and should be fixed with https://github.com/Azure/azure-container-networking/pull/305

**Which issue(s) this PR fixes**:

Fixes #73414 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by PatrickLang** [comment]
/hold
I'm testing this against a pre-release version 1.0.18-rc3 of azure-cni. That needs to be added to the prow scheduled test pass before we merge this.

**Comment by lee0c** [comment]
Removing host entries check in hostname&subdomain test as suggested by @daschott re: this PR: https://github.com/kubernetes/kubernetes/pull/72729

**Comment by michmike** [comment]
We will turn off this test this 1.14 since we are depending on an azurecni and flannel external fixes to complete this work. we will re-enable the tests for 1.15

**Comment by PatrickLang** [comment]
/milestone clear
This will need to merge after 1.14 as we're waiting on CNI updates for these tests to pass. I added this status note to the Windows kep

File: test/e2e/windows/dns.go
Comment (bclau) [inline test/e2e/windows/dns.go]: shouldn't this also work for Linux? If so, then this test should be added to test/e2e/network/dns.go instead, and ideally, the test would also work on Linux.

File: test/e2e/windows/dns.go
Comment (bclau) [inline test/e2e/windows/dns.go]: 2019? :)

File: test/e2e/framework/service_util.go
Comment (lee0c) [reply test/e2e/framework/service_util.go]: Coming into the code base that function was misleading as it doesn't necessarily create a headless service, just allows it to be headless as an option. I can change it back if you still would prefer that but I was attempting to make the naming less misleading.

File: test/e2e/windows/dns.go
Comment (lee0c) [reply test/e2e/windows/dns.go]: Yes, missed that, thank you.

File: test/e2e/windows/dns.go
Comment (lee0c) [reply test/e2e/windows/dns.go]: This is again Windows specific & is in response to #73414

File: test/e2e/windows/dns.go
Comment (lee0c) [reply test/e2e/windows/dns.go]: That test does not pass on windows due to use of dig and other linux-specific commands. I am specifically addressing issue #73414 which asked for Windows compatible DNS tests using Windows-native commands like Resolve-DnsName.

Comment (bclau) [reply test/e2e/windows/dns.go]: This test is using the dnsutils and jessie-dnsutils images, which have dig incorporated in them. They are currently being used by other tests, including other DNS tests which are passing on Windows.

From what I can see, the 2 tests really are testing the same thing, and eventually, the one in test/e2e/network/dns.go will become Conformance sometime in the future and we will have to pass it too, so I don't see why we have to have 2 different versions of the same test.

Comment (lee0c) [reply test/e2e/windows/dns.go]: I have been looking into this & it looks like this test is unnecessary - it fails due to the issue you mentioned [in this PR](https://github.com/kubernetes/kubernetes/pull/72729) & not due to lack of Windows compatibility like we originally thought. I will pull this test from the file as it a duplicate.

File: test/e2e/windows/dns.go
Comment (lee0c) [reply test/e2e/windows/dns.go]: Update, while the other test seems to be a duplicate, I don't believe this will work for Linux as it uses ipconfig instead of ifconfig. It is replacing a test that checks /etc/resolv.conf in linux containers

**Comment by lee0c** [comment]
@bclau Let me know if you want more changes, I removed the redundant test. I can additionally remove the changes to service_util.go if you'd still prefer that.

**Comment by michmike** [comment]
i will let @bclau approve/lgtm and remove the hold once he does his review
thanks

File: test/e2e/framework/service_util.go
Comment (bclau) [reply test/e2e/framework/service_util.go]: Sure, makes sense.

File: test/e2e/windows/dns.go
Comment (bclau) [reply test/e2e/windows/dns.go]: Out of curiosity, which is the ``/etc/resolv.conf`` test you're mentioning?

**Comment by bclau** [comment]
Hmm. The PR mentioned in the commit merged, but it is not tagged at this moment. Ideally, we would have to wait for a new azure-vnet-cni release and make sure it will be used by aks-engine, so we won't end up with a failing test right before the code freeze.

File: test/e2e/windows/dns.go
Comment (lee0c) [reply test/e2e/windows/dns.go]: [This test](https://github.com/kubernetes/kubernetes/blob/18cc11566f25b11c9fb1f2157a85bb40b410c8a7/test/e2e/network/dns.go#L316) which the "should support configurable pod DNS servers" is based on.

**Comment by lee0c** [comment]
> ...we would have to wait for a new azure-vnet-cni release and make sure it will be used by aks-engine, so we won't end up with a failing test right before the code freeze.

I believe Patrick decided this should not be merged until after 1.14 anyway due to waiting on CNI fixes

**Comment by lee0c** [comment]
Additionally, I am going to re-add the [WIP] tag - Patrick tested this with the most updated version of Azure CNI and I missed something that is linux specific. Working this out w/ him now.

**Comment by lee0c** [comment]
@PatrickLang ran this w/ updated Azure CNI & it is passing so removing WIP label

**Comment by michmike** [comment]
Does this mean we should merge as soon as v1.14 is forked? Or is the latest version of azure CNI not merged in yet

**Comment by bclau** [comment]
Should have thought about this before, but we should also make sure that these new tests pass in GCE, vsphere, and flannel as well, in order to make sure we won't have any surprises. I'll check it on flannel l2bridge.

/cc @benmoss

**Comment by benmoss** [comment]
It doesn't work with flannel l2bridge. Looks like we need to wait for https://github.com/containernetworking/plugins/pull/271 to be merged

**Comment by benmoss** [comment]
I would say let's hold off on merging this until it works on more CNIs than just Azure

**Comment by dineshgovindasamy** [comment]
@benmoss https://github.com/containernetworking/plugins/pull/271 is merged. can you try it with Flannel and L2bridge cni. if its working i would like to get this merged.

**Comment by benmoss** [comment]
It looks like the Flannel CNI doesn't support passing `runtimeConfig` to the delegate plugin: https://github.com/containernetworking/plugins/issues/305

**Comment by nagiesek** [comment]
https://github.com/containernetworking/plugins/pull/309

**Comment by timmycarr** [comment]
Hi! Friendly reminder from your release team: we are starting the code freeze for 1.15 tomorrow EOD. Just checking in to see if this is still planned for the 1.15 cycle?

**Comment by lee0c** [comment]
@benmoss Are there more CNIs this is waiting on or is the PR to flannel all that it needed?

**Comment by benmoss** [comment]
I don't know if other CNIs support `runtimeConfig`. Does anyone have a definitive list of all the Windows-capable CNIs? It seems counter to the whole idea of CNI to have this loose dependence on an optional feature, and even more problematic if we are depending on that for defining functioning Windows clusters. However this seems like the decision that for good or bad was accepted in #67435.

I'd kind of like to see us formalize this requirement and elevate this feature to an actual mandatory part of the CNI spec, maybe giving us a way to standardize across Linux and Windows. Linux implementations currently don't deal `/etc/resolv.conf` at all today afaik.

As much as I hate to bike shed this, I do actually care about making sure we do the right thing here. Since this is only a test as well it doesn't seem like it's going to actually be blocking users.
 
/hold
/cc @fasaxc @kubernetes/sig-network-misc

**Comment by thockin** [comment]
Maybe I am missing what part of this PR is related to CNI?

As far as I understand, the "normal" way of setting up DNS (resolv.conf) doesn't work AT ALL for Windows, and the CNI functionality was really the only choice.

That said, I think you are getting at something that has been increasingly bothersome to me - the network "plugin" model on k8s is incredibly incomplete.  I've written about this a bit before, but if it is something I can get people to pursue, I'd be happy to write more.  In short, the network functionality is currently spread between:
* CNI (pod network interfaces, IPAM, HostPorts, now DNS)
* Kubelet (HostPorts, DNS config, hosts file)
* kube-proxy (Service VIPs, NodePorts)
* NetworkPolicy implementations
* controller manager (node IPAM, routes)

What we should do is build a more complete abstraction and excise the network-related logic from these components.  Kubelet/CRI should call out to a plugin that sets up per-pod interfaces AND services AND policies.  IPAM should be distinct controllers.


WRT this PR, I think it's established that this is how Windows works for now.  I'm not sure if that answers your concern.

**Comment by benmoss** [comment]
It's not that this PR that is related to CNI so much, but it's adding a test for functionality that only works *if* you're using a CNI that happens to support that optional feature via the `runtimeConfig` / `capabilities` parts of the CNI spec (which is basically CNI's "here be dragons"). The decision to rely on that was technically made in #67435 but I don't know if the ramifications were fully thought through.

I agree with you that networking details is spread wide, and I guess this is just another (particularly bad imo) case of that. On a pragmatic level it all works, it just seems very kludgy and underspecified.

/hold cancel

**Review by michmike — APPROVED** [review]

File: test/e2e/framework/service_util.go
Comment (BenTheElder) [reply test/e2e/framework/service_util.go]: in general I'd probably recommend doing this sort of thing in a different PR to avoid unnecessarily holding up approvals on the rest of the changes

**Review by BenTheElder — APPROVED** [review]
Approve for the trivial framework change
/approve

File: test/e2e/framework/service_util.go
Comment (lee0c) [reply test/e2e/framework/service_util.go]: I'll keep that in mind in the future 👍

**Comment by benmoss** [comment]
/hold 

@lee0c looks like the `pull-kubernetes-verify` failures are real and not just a flake, can you take a look?

**Comment by lee0c** [comment]
@benmoss Will do, though I haven't looked at this in months & my schedule is pretty tight at the moment so I may be slow to fix. Wouldn't be surprised if this is just due to the branch being out of date at this point.

**Comment by lee0c** [comment]
Rebased & there are still failures. I will have to dig into this soon.

**Comment by PatrickLang** [comment]
@lee0c let me know if you need help on this. I can look next week if needed

**Comment by makoscafee** [comment]
Hello! @lee0c & @PatrickLang  I'm a bug triage shadow for the 1.16 release cycle  I'd like to check it's status. The code freeze is starting 29th August (about 2 weeks from now) and while there is plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the PR is tagged for 1.16, is it still planned for this release?

**Comment by PatrickLang** [comment]
@lee0c - the build breaks are from a refactoring in logging. Here's an example of what needs to change. It should be just a few lines
  https://github.com/kubernetes/kubernetes/commit/332be4b1e37e73b889fd20d7c190e54a34b52bd0

**Review by michmike — APPROVED** [review]
/approve
/lgtm

**Comment by lee0c** [comment]
Thank you so much for your help @PatrickLang !!!

**Pull request merged**
